accord-core/src/main/java/accord/coordinate/CoordinatePreAccept.java

```diff
@@ -152,7 +152,12 @@ void onNewEpochTopologyMismatch(TopologyMismatch mismatch)
     void onPreAccepted(Topologies topologies)
     {
         Timestamp executeAt = foldl(oks, (ok, prev) -> mergeMax(ok.witnessedAt, prev), Timestamp.NONE);
-        node.withEpoch(executeAt.epoch(), () -> onPreAccepted(topologies, executeAt, oks));
+        node.withEpoch(executeAt.epoch(), (ignored, withEpochFailure) -> {
```

Comment:
makes me wonder if `AsyncChain` is the better solution.  What happens if the success path throws? In this case `onPreAccepted` throws then we don't have any way to catch/trap that exception

Reply:
The issue with callbacks is that failures bubble up in random places, where as `AsyncChain` kinda forces `begin` to handle everything.  Right now if the epoch is known this call trace must handle the exceptions, and if not then we send to the `agent` (as it becomes a `AsyncChain` under the hood).

Reply:
another advantage is that you can focus on the happy-path when building the chain and only focus on the failure in one place.  With callbacks we keep having to focus on both cases in every single block...

Reply:
Gave this example on slack...

`accord-core/src/main/java/accord/coordinate/Recover.java`

```
node.withEpoch(invalidateUntil.epoch(), (ignored, withEpochFailure) -> {
            // TODO (review): There was already nothing here to handle failures in commitInvalidate
            if (withEpochFailure != null)
            {
                node.agent().onUncaughtException(CoordinationFailed.wrap(withEpochFailure));
                return;
            }
            Commit.Invalidate.commitInvalidate(node, txnId, route, invalidateUntil);
        });
```

vs

```
node.withEpoch(invalidateUntil.epoch())
    .accept(ignore -> Commit.Invalidate.commitInvalidate(node, txnId, route, invalidateUntil))
    .begin(agent);
```

For this project, when given the success and failure we keep ignoring the errors, and this patch is kinda trying to address that by handling the errors... but we keep having this problem over and over again; callbacks have been very error prone for us...

By using `AsyncChain` we can focus on the success cases only and can send the unexpected / unhandled errors to the `agent`